src/ansi.rs

```diff
@@ -672,6 +673,70 @@ pub trait Handler {
 
     /// Set mouse cursor icon.
     fn set_mouse_cursor_icon(&mut self, _: CursorIcon) {}
+
+    /// Report current keyboard mode.
+    ///
+    /// ## Caveat
+    ///
+    /// This request must be done successively with the [`identify_terminal`] request, given
+    /// clients rely on the later to identify whether the terminal suppors kitty's keyboard
+    /// protocol.
+    ///
+    /// [`identify_terminal`]: Self::identify_terminal
+    fn report_keyboard_mode(&mut self) {}
+
+    /// Push keyboard mode into the keyboard mode stack.
+    fn push_keyboard_mode(&mut self, _mode: KeyboardMode) {}
+
+    /// Pop the given amount of keyboard modes from the
+    /// keyboard mode stack.
+    fn pop_num_keyboard_modes(&mut self, _to_pop: u16) {}
+
+    /// Set the [`keyboard mode`] using the given [`behavior`].
+    ///
+    /// [`keyboard mode`]: crate::ansi::KeyboardMode
+    /// [`behavior`]: crate::ansi::KeyboardModeApplyBehavior
+    fn set_keyboard_mode(&mut self, _mode: KeyboardMode, _behavior: KeyboardModeApplyBehavior) {}
+}
+
+bitflags! {
+    /// The keyboard mode representing [`kitty keyboard protocol'].
+    ///
+    /// The details on how to build the escape sequences could be found in the
+    /// mentioned specification.
+    ///
+    /// [`kitty keyboard protocol']: https://sw.kovidgoyal.net/kitty/keyboard-protocol
+    #[derive(Default, Debug, Clone, Copy, PartialEq, Eq, Hash)]
+    pub struct KeyboardMode : u8 {
+        /// No particular mode is being set.
+        const NO_MODE                 = 0b0000_0000;
+        /// Report `Esc`, `alt` + `key`, `ctrl` + `key`, `ctrl` + `alt` + `key`, `shift`
+        /// + `alt` + `key` keys using `CSI u` sequence instead of legacy ones.
+        const DISAMBIGUATE_ESC_CODES  = 0b0000_0001;
+        /// Report key presses, release, repeat along side the escape. If the
+        /// key event results in text the type is not support for it, unless
+        /// [`REPORT_ALL_KEYS_AS_ESC`] is enabled.
+        const REPORT_EVENT_TYPES      = 0b0000_0010;
+        /// Report alternative value in pair with original value.
+        const REPORT_ALTERNATE_KEYS   = 0b0000_0100;
+        /// Report every key as an escape sequence.
+        const REPORT_ALL_KEYS_AS_ESC  = 0b0000_1000;
+        /// Report the text generated by the key event.
+        const REPORT_ASSOCIATED_TEXT  = 0b0001_0000;
+    }
+}
+
+/// Describes how the new [`KeyboardMode`] should be applied.
+#[repr(u8)]
+#[derive(Default, Clone, Copy, PartialEq, Eq)]
+pub enum KeyboardModeApplyBehavior {
```

Comment:
Generally I'd say "ApplicationBehavior" is more correct here, but I think it's better to leave it as "Apply" since application has multiple meanings.